pkg/cloud/aws/actuators/machine/actuator.go

```diff
@@ -241,6 +241,11 @@ func (a *Actuator) Delete(ctx context.Context, cluster *clusterv1.Cluster, machi
 
 	ec2svc := ec2.NewService(scope.Scope)
 
+	if scope.MachineStatus.InstanceID == nil {
+		klog.Error("Instance does not have an instance id")
+		return nil
```

Comment:
What does it mean when `instanceId` is nil?
If `instanceId` is nil, then that means that the machine object hasn't been reconciled yet or is in the process of being reconciled by another thread. So there could be a race. For that reason, I think we shouldn't be returning nil (no error) here. By returning an error, this delete will be re-queued and will be processed once the reconciliation is complete.

Reply:
This is trying to capture the case where the `instanceID` is nil because the ec2 instance never existed (wrong iam role or any other issue). The retries would happen over and over again but with no change. In that case, it seems that the reconciliation process wouldn't make any difference.

Reply:
There is another use case that I hit as well where the instance exists but the Status has not been fully populated yet (pivoting the cluster resources). Handling this in InstanceIfExists would allow for the filtered lookup by labels to resolve both issues.

Reply:
@yozamacs if we were to return `nil` here, what would happen if the machine is in the process of reconciliation?

Reply:
I'm not sure. Where does the reconciliation happen? I'm not very familiar with that process

Reply:
Answered here https://kubernetes.slack.com/archives/CD6U2V71N/p1552434137136400?thread_ts=1552423306.129400&cid=CD6U2V71N.
returning nil error should be fine :)